File: crack.c
#define _XOPEN_SOURCE
#include <unistd.h>
#include <stdio.h>
#include <string.h>

int main(int argc, char const *argv[])
{
    // condition to run
    if (argc == 2)
    {
        // print input hash
        const char *user_input = argv[1];

        // create an array of alhpabet letters
        const char alphabet[53] = {' ', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'};

        // create salt
        char salt[3];
        salt[0] = user_input[0];
        salt[1] = user_input[1];
        salt[2] = '\0';

        // decifre input hash
        char password[6];

        for (int i = 0; i < strlen(alphabet); i++) //for 5 letter passwords
        {
            if (i > 0)
            {
                password[5] = '\0';
                password[4] = alphabet[i];
            }

            for (int j = 0; j < strlen(alphabet); j++) //for 4 letter passwords
            {
                if (i == 0)
                {
                    password[4] = '\0';
                }

                password[3] = alphabet[j];

                for (int k = 0; k < strlen(alphabet); k++) // for 3 letter password
                {
                    if (j == 0)
                    {
                        password[3] = '\0';
                    }

                    password[2] = alphabet[k];

                    for (int l = 0; l < strlen(alphabet); l++) // for 2 letter password
                    {
                        if (k == 0)
                        {
                            password[2] = '\0';
                        }

                        password[1] = alphabet[l];

                        for (int m = 0; m < strlen(alphabet); m++) // for 1 letter password
                        {
                            if (l == 0)
                            {
                                password[1] = '\0';
                            }

                            password[0] = alphabet[m];

                            if (strcmp(crypt(password, salt), argv[1]) == 0)
                            {
                                printf("%s\n", password);
                                return 0;
                            }
                        }
                    }
                }
            }
        }
    }

    // no or too many arguments
    else
    {
        printf("Usage: ./crack hash\n");
        return 1;
    }
}

File: crack.py
import sys
import crypt
import string

# function returnes salt


def get_salt():

    hash_val = sys.argv[1]
    salt = ""

    for letter in hash_val[0:2:]:
        salt += letter

    return salt


# check that there are onyl 2 command line arguments
def main():
    if len(sys.argv) != 2:

        # print usage message and exit with code 1
        print("Usage: ./crack hash")
        exit(1)

    else:

        # get the salt
        salt = get_salt()

        # iterate through 1 letter password
        for i in string.ascii_letters:

            password = ""
            password += i

            if crypt.crypt(password, salt) == sys.argv[1]:

                print(password)
                exit(0)

        print("Pregledu 1")

        # iterate through 2 letter password
        for i in string.ascii_letters:

            password = ""
            password += i

            if crypt.crypt(password, salt) == sys.argv[1]:

                print(password)

            else:
                for i in string.ascii_letters:

                    password_1 = password
                    password_1 += i

                    if crypt.crypt(password_1, salt) == sys.argv[1]:

                        print(password_1)
                        exit(0)

        print("Pregledu 2")

        # iterate through 3 letter password
        for i in string.ascii_letters:

            password = ""
            password += i

            if crypt.crypt(password, salt) == sys.argv[1]:

                print(password)

            else:
                for j in string.ascii_letters:

                    password_1 = password
                    password_1 += j

                    if crypt.crypt(password_1, salt) == sys.argv[1]:

                        print(password_1)

                    else:
                        for k in string.ascii_letters:

                            password_2 = password_1
                            password_2 += k

                            if crypt.crypt(password_2, salt) == sys.argv[1]:

                                print(password_2)
                                exit(0)

        print("Pregledu 3")

        # iterate through 4 letter password
        for i in string.ascii_letters:

            password = ""
            password += i

            if crypt.crypt(password, salt) == sys.argv[1]:

                print(password)

            else:
                for j in string.ascii_letters:

                    password_1 = password
                    password_1 += j

                    if crypt.crypt(password_1, salt) == sys.argv[1]:

                        print(password_1)

                    else:
                        for k in string.ascii_letters:

                            password_2 = password_1
                            password_2 += k

                            if crypt.crypt(password_2, salt) == sys.argv[1]:

                                print(password_2)

                            else:
                                for l in string.ascii_letters:

                                    password_3 = password_2
                                    password_3 += l

                                    if crypt.crypt(password_3, salt) == sys.argv[1]:

                                        print(password_3)
                                        exit(0)

        print("Pregledu 4")

        # iterate through 5 letter password
        for i in string.ascii_letters:

            password = ""
            password += i

            if crypt.crypt(password, salt) == sys.argv[1]:

                print(password)

            else:
                for j in string.ascii_letters:

                    password_1 = password
                    password_1 += j

                    print(f"Zdele sem pri črki {password_1}")

                    if crypt.crypt(password_1, salt) == sys.argv[1]:

                        print(password_1)

                    else:
                        for k in string.ascii_letters:

                            password_2 = password_1
                            password_2 += k

                            if crypt.crypt(password_2, salt) == sys.argv[1]:

                                print(password_2)

                            else:
                                for l in string.ascii_letters:

                                    password_3 = password_2
                                    password_3 += l

                                    if crypt.crypt(password_3, salt) == sys.argv[1]:

                                        print(password_3)

                                    else:
                                        for m in string.ascii_letters:

                                            password_4 = password_3
                                            password_4 += m

                                            if crypt.crypt(password_4, salt) == sys.argv[1]:

                                                print(password_4)
                                                exit(0)


# call main function
if __name__ == "__main__":
    main()